File: ngs/ngs-java/ngs/itf/ReferenceItf.java
/*===========================================================================
*
*                            PUBLIC DOMAIN NOTICE
*               National Center for Biotechnology Information
*
*  This software/database is a "United States Government Work" under the
*  terms of the United States Copyright Act.  It was written as part of
*  the author's official duties as a United States Government employee and
*  thus cannot be copyrighted.  This software/database is freely available
*  to the public for use. The National Library of Medicine and the U.S.
*  Government have not placed any restriction on its use or reproduction.
*
*  Although all reasonable efforts have been taken to ensure the accuracy
*  and reliability of the software and data, the NLM and the U.S.
*  Government do not and cannot warrant the performance or results that
*  may be obtained by using this software or data. The NLM and the U.S.
*  Government disclaim all warranties, express or implied, including
*  warranties of performance, merchantability or fitness for any particular
*  purpose.
*
*  Please cite the author in any work or product based on this material.
*
* ===========================================================================
*
*/

package ngs.itf;

import ngs.ErrorMsg;
import ngs.Reference;
import ngs.Alignment;
import ngs.AlignmentIterator;
import ngs.PileupIterator;


/*==========================================================================
 * ReferenceItf
 *  represents a reference sequence
 */
class ReferenceItf
    extends Refcount
    implements Reference
{

    /***********************
     * Reference Interface *
     ***********************/

    /* getCommonName
     *  returns the common name of reference, e.g. "chr1"
     */
    public String getCommonName ()
        throws ErrorMsg
    {
        return this . GetCommonName ( self );
    }

    /* getCanonicalName
     *  returns the accessioned name of reference, e.g. "NC_000001.11"
     */
    public String getCanonicalName ()
        throws ErrorMsg
    {
        return this . GetCanonicalName ( self );
    }


    /* getIsCircular
     *  returns true if reference is circular
     */
    public boolean getIsCircular ()
        throws ErrorMsg
    {
        return this . GetIsCircular ( self );
    }


    /* getIsLocal
     *  returns true if reference is local
     */
    public boolean getIsLocal ()
        throws ErrorMsg
    {
        return this . GetIsLocal ( self );
    }


    /* getLength
     *  returns the length of the reference sequence
     */
    public long getLength ()
        throws ErrorMsg
    {
        return this . GetLength ( self );
    }


    /* getReferenceBases
     *  return sub-sequence bases for Reference
     *  "offset" is zero-based
     *  "size" must be >= 0
     */
    public String getReferenceBases ( long offset )
        throws ErrorMsg
    {
        return this . getReferenceBases ( offset, -1 );
    }

    public String getReferenceBases ( long offset, long length )
        throws ErrorMsg
    {
        return this . GetReferenceBases ( self, offset, length );
    }

    /* getReferenceChunk
     *  return largest contiguous chunk available of
     *  sub-sequence bases for Reference
     *  "offset" is zero-based
     *  "size" must be >= 0
     *
     * NB - actual returned sequence may be shorter
     *  than requested. to obtain all bases available
     *  in chunk, use a negative "size" value
     */
    public String getReferenceChunk ( long offset )
        throws ErrorMsg
    {
        return this . getReferenceChunk ( offset, -1 );
    }

    public String getReferenceChunk ( long offset, long length )
        throws ErrorMsg
    {
        return this . GetReferenceChunk ( self, offset, length );
    }


    /*----------------------------------------------------------------------
     * ALIGNMENTS
     */

    /* getAlignmentCount
     */
    public long getAlignmentCount ()
        throws ErrorMsg
    {
        return this . GetAlignmentCount ( self, Alignment . all );
    }

    public long getAlignmentCount ( int categories )
        throws ErrorMsg
    {
        return this . GetAlignmentCount ( self, categories );
    }

    /* getAlignment
     *  returns an individual Alignment
     *  throws ErrorMsg if Alignment does not exist
     *  or is not part of this Reference
     */
    public Alignment getAlignment ( String alignmentId )
        throws ErrorMsg
    {
        long ref = this . GetAlignment ( self, alignmentId );
        try
        {
            return new AlignmentItf ( ref );
        }
        catch ( Exception x )
        {
            this . release ( ref );
            throw new ErrorMsg ( x . toString () );
        }
    }

    /* AlignmentCategory
     *  see Alignment for categories
     */

    /* getAlignments
     *  returns an iterator of contained alignments
     */
    public AlignmentIterator getAlignments ( int categories )
        throws ErrorMsg
    {
        long ref = this . GetAlignments ( self, categories );
        try
        {
            return new AlignmentIteratorItf ( ref );
        }
        catch ( Exception x )
        {
            this . release ( ref );
            throw new ErrorMsg ( x . toString () );
        }
    }

    /* getAlignmentSlice
     *  returns an iterator across a range of Alignments
     *  "first" is an unsigned ordinal into set
     *  "categories" provides a means of filtering by AlignmentCategory
     */
    public AlignmentIterator getAlignmentSlice ( long offset, long length )
        throws ErrorMsg
    {
        return this . getAlignmentSlice ( offset, length, Alignment . all );
    }

    public AlignmentIterator getAlignmentSlice ( long offset, long length, int categories )
        throws ErrorMsg
    {
        long ref = this . GetAlignmentSlice ( self, offset, length, categories );
        try
        {
            return new AlignmentIteratorItf ( ref );
        }
        catch ( Exception x )
        {
            this . release ( ref );
            throw new ErrorMsg ( x . toString () );
        }
    }

    /* getFilteredAlignmentSlice
     *  returns a filtered iterator across a slice of the Reference
     *  behaves like "getAlignmentSlice" except that supported filters are applied to selection
     *  "filters" is a set of filter bits defined in Alignment
     *  "mappingQuality" is a cutoff to be used according to bits in "filters"
     */
    public AlignmentIterator getFilteredAlignmentSlice ( long offset, long length, int categories, int filters, int mappingQuality )
        throws ErrorMsg
    {
        long ref = this . GetFilteredAlignmentSlice ( self, offset, length, categories, filters, mappingQuality );
        try
        {
            return new AlignmentIteratorItf ( ref );
        }
        catch ( Exception x )
        {
            this . release ( ref );
            throw new ErrorMsg ( x . toString () );
        }
    }


    /*----------------------------------------------------------------------
     * PILEUP
     */

    /* getPileups
     *  returns an iterator of contained Pileups
     */
    public PileupIterator getPileups ( int categories )
        throws ErrorMsg
    {
        long ref = this . GetPileups ( self, categories );
        try
        {
            return new PileupIteratorItf ( ref );
        }
        catch ( Exception x )
        {
            this . release ( ref );
            throw new ErrorMsg ( x . toString () );
        }
    }

    public PileupIterator getFilteredPileups ( int categories, int filters, int mappingQuality )
        throws ErrorMsg
    {
        long ref = this . GetFilteredPileups ( self, categories, filters, mappingQuality );
        try
        {
            return new PileupIteratorItf ( ref );
        }
        catch ( Exception x )
        {
            this . release ( ref );
            throw new ErrorMsg ( x . toString () );
        }
    }

    /* getPileupSlice
     *  creates a PileupIterator on a slice (window) of reference
     *  "start" is the signed starting position on reference
     *  "length" is the unsigned number of bases in the window
     *  "categories" provides a means of filtering by AlignmentCategory
     */
    public PileupIterator getPileupSlice ( long offset, long length )
        throws ErrorMsg
    {
        return this . getPileupSlice ( offset, length, Alignment . all );
    }

    public PileupIterator getPileupSlice ( long offset, long length, int categories )
        throws ErrorMsg
    {
        long ref = this . GetPileupSlice ( self, offset, length, categories );
        try
        {
            return new PileupIteratorItf ( ref );
        }
        catch ( Exception x )
        {
            this . release ( ref );
            throw new ErrorMsg ( x . toString () );
        }
    }

    public PileupIterator getFilteredPileupSlice ( long offset, long length, int categories, int filters, int mappingQuality )
        throws ErrorMsg
    {
        long ref = this . GetFilteredPileupSlice ( self, offset, length, categories, filters, mappingQuality );
        try
        {
            return new PileupIteratorItf ( ref );
        }
        catch ( Exception x )
        {
            this . release ( ref );
            throw new ErrorMsg ( x . toString () );
        }
    }

    /*******************************
     * ReferenceItf Implementation *
     *******************************/

    // constructors
    ReferenceItf ( long ref )
    {
        super ( ref );
    }

    ReferenceItf ( Reference obj )
        throws ErrorMsg
    {
        super ( 0 );
        try
        {
            ReferenceItf ref = ( ReferenceItf ) obj;
            this . self = ref . duplicate ();
        }
        catch ( Exception x )
        {
            throw new ErrorMsg ( x . toString () );
        }
    }

    // native interface
    private native String GetCommonName ( long self )
        throws ErrorMsg;
    private native String GetCanonicalName ( long self )
        throws ErrorMsg;
        private native boolean GetIsCircular ( long self )
        throws ErrorMsg;
    private native boolean GetIsLocal ( long self )
        throws ErrorMsg;
    private native long GetLength ( long self )
        throws ErrorMsg;
    private native String GetReferenceBases ( long self, long offset, long length )
        throws ErrorMsg;
    private native String GetReferenceChunk ( long self, long offset, long length )
        throws ErrorMsg;
    private native long GetAlignmentCount ( long self, int categories )
        throws ErrorMsg;
    private native long GetAlignment ( long self, String alignmentId )
        throws ErrorMsg;
    private native long GetAlignments ( long self, int categories )
        throws ErrorMsg;
    private native long GetAlignmentSlice ( long self, long offset, long length, int categories )
        throws ErrorMsg;
    private native long GetFilteredAlignmentSlice ( long self, long offset, long length, int categories, int filters, int mappingQuality )
        throws ErrorMsg;
    private native long GetPileups ( long self, int categories )
        throws ErrorMsg;
    private native long GetFilteredPileups ( long self, int categories, int filters, int mappingQuality )
        throws ErrorMsg;
    private native long GetPileupSlice ( long self, long offset, long count, int categories )
        throws ErrorMsg;
    private native long GetFilteredPileupSlice ( long self, long offset, long count, int categories, int filters, int mappingQuality )
        throws ErrorMsg;
}

File: ngs/ngs-sdk/adapter/ReferenceItf.cpp
/*===========================================================================
*
*                            PUBLIC DOMAIN NOTICE
*               National Center for Biotechnology Information
*
*  This software/database is a "United States Government Work" under the
*  terms of the United States Copyright Act.  It was written as part of
*  the author's official duties as a United States Government employee and
*  thus cannot be copyrighted.  This software/database is freely available
*  to the public for use. The National Library of Medicine and the U.S.
*  Government have not placed any restriction on its use or reproduction.
*
*  Although all reasonable efforts have been taken to ensure the accuracy
*  and reliability of the software and data, the NLM and the U.S.
*  Government do not and cannot warrant the performance or results that
*  may be obtained by using this software or data. The NLM and the U.S.
*  Government disclaim all warranties, express or implied, including
*  warranties of performance, merchantability or fitness for any particular
*  purpose.
*
*  Please cite the author in any work or product based on this material.
*
* ===========================================================================
*
*/

#include <ngs/adapter/ReferenceItf.hpp>
#include <ngs/adapter/StringItf.hpp>
#include <ngs/adapter/StatisticsItf.hpp>
#include <ngs/adapter/AlignmentItf.hpp>
#include <ngs/adapter/PileupItf.hpp>
#include <ngs/adapter/ErrorMsg.hpp>

#include "ErrBlock.hpp"

namespace ngs_adapt
{

    /*----------------------------------------------------------------------
     * ReferenceItf
     */

    ReferenceItf :: ReferenceItf ()
        : Refcount < ReferenceItf, NGS_Reference_v1 > ( & ivt . dad )
    {
    }

    NGS_String_v1 * CC ReferenceItf :: get_cmn_name ( const NGS_Reference_v1 * iself, NGS_ErrBlock_v1 * err )
    {
        const ReferenceItf * self = Self ( iself );
        try
        {
            StringItf * val = self -> getCommonName ();
            return val -> Cast ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return 0;
    }

    NGS_String_v1 * CC ReferenceItf :: get_canon_name ( const NGS_Reference_v1 * iself, NGS_ErrBlock_v1 * err )
    {
        const ReferenceItf * self = Self ( iself );
        try
        {
            StringItf * val = self -> getCanonicalName ();
            return val -> Cast ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return 0;
    }

    bool CC ReferenceItf :: is_circular ( const NGS_Reference_v1 * iself, NGS_ErrBlock_v1 * err )
    {
        const ReferenceItf * self = Self ( iself );
        try
        {
            return self -> getIsCircular ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return 0;
    }

    bool CC ReferenceItf :: is_local ( const NGS_Reference_v1 * iself, NGS_ErrBlock_v1 * err )
    {
        const ReferenceItf * self = Self ( iself );
        try
        {
            return self -> getIsLocal ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return 0;
    }

    uint64_t CC ReferenceItf :: get_length ( const NGS_Reference_v1 * iself, NGS_ErrBlock_v1 * err )
    {
        const ReferenceItf * self = Self ( iself );
        try
        {
            return self -> getLength ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return 0;
    }

    NGS_String_v1 * CC ReferenceItf :: get_ref_bases ( const NGS_Reference_v1 * iself, NGS_ErrBlock_v1 * err,
            uint64_t offset, uint64_t length )
    {
        const ReferenceItf * self = Self ( iself );
        try
        {
            StringItf * val = self -> getReferenceBases ( offset, length );
            return val -> Cast ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return 0;
    }

    NGS_String_v1 * CC ReferenceItf :: get_ref_chunk ( const NGS_Reference_v1 * iself, NGS_ErrBlock_v1 * err,
            uint64_t offset, uint64_t length )
    {
        const ReferenceItf * self = Self ( iself );
        try
        {
            StringItf * val = self -> getReferenceChunk ( offset, length );
            return val -> Cast ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return 0;
    }

    uint64_t CC ReferenceItf :: get_alignment_count ( const NGS_Reference_v1 * iself, NGS_ErrBlock_v1 * err,
            bool wants_primary, bool wants_secondary )
    {
        const ReferenceItf * self = Self ( iself );
        try
        {
            return self -> getAlignmentCount ( wants_primary, wants_secondary );
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return 0;
    }

    NGS_Alignment_v1 * CC ReferenceItf :: get_alignment ( const NGS_Reference_v1 * iself, NGS_ErrBlock_v1 * err,
            const char * alignmentId )
    {
        const ReferenceItf * self = Self ( iself );
        try
        {
            AlignmentItf * val = self -> getAlignment ( alignmentId );
            return val -> Cast ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return 0;
    }

    NGS_Alignment_v1 * CC ReferenceItf :: get_alignments ( const NGS_Reference_v1 * iself, NGS_ErrBlock_v1 * err,
                                           bool wants_primary, bool wants_secondary )
    {
        const ReferenceItf * self = Self ( iself );
        try
        {
            AlignmentItf * val = self -> getAlignments ( wants_primary, wants_secondary );
            return val -> Cast ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return 0;
    }

    NGS_Alignment_v1 * CC ReferenceItf :: get_align_slice ( const NGS_Reference_v1 * iself, NGS_ErrBlock_v1 * err,
                                            int64_t start, uint64_t length, bool wants_primary, bool wants_secondary )
    {
        const ReferenceItf * self = Self ( iself );
        try
        {
            AlignmentItf * val = self -> getAlignmentSlice ( start, length, wants_primary, wants_secondary );
            return val -> Cast ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return 0;
    }

    NGS_Alignment_v1 * CC ReferenceItf :: get_filtered_align_slice ( const NGS_Reference_v1 * iself, NGS_ErrBlock_v1 * err,
                                            int64_t start, uint64_t length, uint32_t flags, int32_t map_qual )
    {
        const ReferenceItf * self = Self ( iself );
        try
        {
            AlignmentItf * val = self -> getFilteredAlignmentSlice ( start, length, flags, map_qual );
            return val -> Cast ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return 0;
    }

    NGS_Pileup_v1 * CC ReferenceItf :: get_pileups ( const NGS_Reference_v1 * iself, NGS_ErrBlock_v1 * err,
                                     bool wants_primary, bool wants_secondary )
    {
        const ReferenceItf * self = Self ( iself );
        try
        {
            PileupItf * val = self -> getPileups ( wants_primary, wants_secondary );
            return val -> Cast ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return 0;
    }

    NGS_Pileup_v1 * CC ReferenceItf :: get_filtered_pileups ( const NGS_Reference_v1 * iself, NGS_ErrBlock_v1 * err,
        uint32_t flags, int32_t map_qual )
    {
        const ReferenceItf * self = Self ( iself );
        try
        {
            PileupItf * val = self -> getFilteredPileups ( flags, map_qual );
            return val -> Cast ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return 0;
    }

    NGS_Pileup_v1 * CC ReferenceItf :: get_pileup_slice ( const NGS_Reference_v1 * iself, NGS_ErrBlock_v1 * err,
                                          int64_t start, uint64_t length, bool wants_primary, bool wants_secondary )
    {
        const ReferenceItf * self = Self ( iself );
        try
        {
            PileupItf * val = self -> getPileupSlice ( start, length, wants_primary, wants_secondary );
            return val -> Cast ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return 0;
    }

    NGS_Pileup_v1 * CC ReferenceItf :: get_filtered_pileup_slice ( const NGS_Reference_v1 * iself, NGS_ErrBlock_v1 * err,
        int64_t start, uint64_t length, uint32_t flags, int32_t map_qual )
    {
        const ReferenceItf * self = Self ( iself );
        try
        {
            PileupItf * val = self -> getFilteredPileupSlice ( start, length, flags, map_qual );
            return val -> Cast ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return 0;
    }

    bool CC ReferenceItf :: next ( NGS_Reference_v1 * iself, NGS_ErrBlock_v1 * err )
    {
        ReferenceItf * self = Self ( iself );
        try
        {
            return self -> nextReference ();
        }
        catch ( ... )
        {
            ErrBlockHandleException ( err );
        }

        return false;
    }

    NGS_Reference_v1_vt ReferenceItf :: ivt =
    {
        {
            "ngs_adapt::ReferenceItf",
            "NGS_Reference_v1",
            4,
            & OpaqueRefcount :: ivt . dad
        },

        // 1.0
        get_cmn_name,
        get_canon_name,
        is_circular,
        get_length,
        get_ref_bases,
        get_ref_chunk,
        get_alignment,
        get_alignments,
        get_align_slice,
        get_pileups,
        get_pileup_slice,
        next,

        // 1.1
        get_filtered_pileups,
        get_filtered_pileup_slice,

        // 1.2
        get_alignment_count,

        // 1.3 interface
        NULL,// get_filtered_alignments (not exposed via ngs::Reference)
        get_filtered_align_slice,

        // 1.4
        is_local
    };

} // namespace ngs_adapt
